python/triton/runtime/interpreter.py

```diff
@@ -1081,22 +1085,34 @@ def _to_cpu(arg):
         for key, value in kwargs.items():
             if hasattr(value, "data_ptr"):
                 kwargs_hst[key] = value.cpu()
+            elif type(value) is tuple:
+                return tuple(map(_to_cpu, value))
             else:
                 kwargs_hst[key] = value
         return args_hst, kwargs_hst
 
     def _restore_args_dev(self, args_dev, args_hst, kwargs, kwargs_hst):
-        for arg_dev, arg_hst in zip(args_dev, args_hst):
+
+        def _from_cpu(arg_dev, arg_hst):
             if hasattr(arg_dev, "data_ptr"):
                 # No need to rewrap because this just modifies internal
                 arg_dev, arg_hst = _unwrap_tensor(arg_dev), _unwrap_tensor(arg_hst)
                 arg_dev.untyped_storage().copy_(arg_hst.untyped_storage())
+            elif type(arg_dev) is tuple:
+                for (arg_dev, arg_hst) in zip(arg_dev, arg_hst):
+                    _from_cpu(arg_dev, arg_hst)
+
+        for arg_dev, arg_hst in zip(args_dev, args_hst):
+            _from_cpu(arg_dev, arg_hst)
 
         # Restore keyword arguments
         for key, kwarg_dev in kwargs.items():
             kwarg_hst = kwargs_hst[key]
             if hasattr(kwarg_dev, "data_ptr"):
                 kwarg_dev.data.copy_(kwarg_hst.to(kwarg_dev.device).data)
```

Comment:
Could it be handled by `_from_cpu` as well?

Reply:
Possibly? I don't quite understand why we have a different path for kwargs anyway. Shouldn't it be operating on wrapped tensors too…?

Reply:
Please address this issue as well.

Reply:
Changed